File: backend/app/services/scanner.py
"""
R2 folder scanner. Builds in-memory AppData from the bucket folder structure
and local scene JSON files.

R2 folder conventions:
  ambiences/{category-slug}/{ambience-slug}.ogg
  ambiences/{category-slug}/cover.webp
  ambiences/{category-slug}/cover.thumb.webp

  playlists/{category-slug}/{playlist-slug}/{track-slug}.ogg
  playlists/{category-slug}/{playlist-slug}/cover.webp
  playlists/{category-slug}/{playlist-slug}/cover.thumb.webp

Scene configs live in backend/data/scenes/ as JSON files (not in R2).
Ambience and playlist slugs (filename stems) are used as stable entity IDs
and must be globally unique within their type.
"""

import json
import logging
import re
from datetime import datetime, timezone
from pathlib import Path

from app.core.storage import r2
from app.schemas import (
    AmbienceCategoryOut,
    AmbienceEntry,
    AmbienceOut,
    BackgroundAssetOut,
    LayerAssetOut,
    MusicTrackOut,
    PlaylistCategoryEntryOut,
    PlaylistCategoryOut,
    PlaylistOut,
    SceneCategoryEntryOut,
    SceneCategoryOut,
    SceneOut,
)
from app.state import AppData

logger = logging.getLogger(__name__)


def _label(slug: str) -> str:
    """Strip leading numeric prefix and convert hyphens to title-cased words."""
    return re.sub(r"^\d+-", "", slug).replace("-", " ").title()


def _order(slug: str) -> int:
    """Return the leading numeric prefix of a slug as a sort key, 999 if absent."""
    m = re.match(r"^(\d+)-", slug)
    return int(m.group(1)) if m else 999


def _scan_ambiences(keys: list[str]) -> tuple[list[AmbienceCategoryOut], list[AmbienceOut]]:
    """Build ambience categories and a flat ambience list from R2 keys."""
    # cat_slug -> cover meta
    cats: dict[str, dict] = {}
    # (cat_slug, amb_slug) -> AmbienceOut
    ambiences: dict[tuple[str, str], AmbienceOut] = {}

    for key in keys:
        parts = key.split("/")
        if len(parts) != 3 or parts[0] != "ambiences":
            continue
        cat_slug, filename = parts[1], parts[2]

        if cat_slug not in cats:
            cats[cat_slug] = {"src": "", "thumb_src": None}

        if filename == "cover.webp":
            cats[cat_slug]["src"] = key
        elif filename == "cover.thumb.webp":
            cats[cat_slug]["thumb_src"] = key
        elif filename.endswith(".ogg"):
            amb_slug = Path(filename).stem
            if (cat_slug, amb_slug) in ambiences:
                logger.warning("Duplicate ambience slug '%s' in category '%s' — skipping", amb_slug, cat_slug)
                continue
            ambiences[(cat_slug, amb_slug)] = AmbienceOut(
                id=amb_slug,
                label=_label(amb_slug),
                src=key,
            )

    result_cats: list[AmbienceCategoryOut] = []
    for cat_slug, meta in sorted(cats.items(), key=lambda x: _order(x[0])):
        entries = [
            AmbienceEntry(id=slug, label=_label(slug))
            for (cs, slug) in sorted(ambiences, key=lambda k: k[1])
            if cs == cat_slug
        ]
        result_cats.append(AmbienceCategoryOut(
            id=cat_slug,
            label=_label(cat_slug),
            src=meta["src"],
            thumb_src=meta["thumb_src"],
            order=_order(cat_slug),
            ambiences=entries,
        ))

    return result_cats, list(ambiences.values())


def _scan_playlists(keys: list[str]) -> tuple[list[PlaylistCategoryOut], list[PlaylistOut]]:
    """Build playlist categories and full playlist objects from R2 keys."""
    # cat_slug -> set (just to track existence)
    cats: dict[str, None] = {}
    # (cat_slug, pl_slug) -> mutable playlist data
    pl_data: dict[tuple[str, str], dict] = {}

    for key in keys:
        parts = key.split("/")
        if len(parts) != 4 or parts[0] != "playlists":
            continue
        cat_slug, pl_slug, filename = parts[1], parts[2], parts[3]

        cats.setdefault(cat_slug, None)
        pl_key = (cat_slug, pl_slug)
        if pl_key not in pl_data:
            pl_data[pl_key] = {"src": "", "thumb_src": None, "tracks": []}

        if filename == "cover.webp":
            pl_data[pl_key]["src"] = key
        elif filename == "cover.thumb.webp":
            pl_data[pl_key]["thumb_src"] = key
        elif filename.endswith(".ogg"):
            track_slug = Path(filename).stem
            pl_data[pl_key]["tracks"].append(MusicTrackOut(id=track_slug, src=key))

    # Sort tracks by src path — numeric filename prefix ensures correct order
    for meta in pl_data.values():
        meta["tracks"].sort(key=lambda t: t.src)

    playlists: list[PlaylistOut] = []
    cat_entries: dict[str, list[PlaylistCategoryEntryOut]] = {s: [] for s in cats}

    for (cat_slug, pl_slug), meta in sorted(pl_data.items()):
        pl = PlaylistOut(
            id=pl_slug,
            label=_label(pl_slug),
            src=meta["src"],
            thumb_src=meta["thumb_src"],
            tracks=meta["tracks"],
        )
        playlists.append(pl)
        cat_entries[cat_slug].append(PlaylistCategoryEntryOut(id=pl_slug, label=_label(pl_slug)))

    result_cats = [
        PlaylistCategoryOut(
            id=cat_slug,
            label=_label(cat_slug),
            order=_order(cat_slug),
            playlists=cat_entries[cat_slug],
        )
        for cat_slug in sorted(cats, key=_order)
    ]

    return result_cats, playlists


def _scan_scenes(scenes_dir: Path) -> tuple[list[SceneCategoryOut], list[SceneOut]]:
    """Read scene JSON configs from disk and build scene + category structures."""
    if not scenes_dir.is_dir():
        return [], []

    scenes: list[SceneOut] = []
    cat_scenes: dict[str, list[SceneCategoryEntryOut]] = {}

    for path in sorted(scenes_dir.glob("*.json")):
        try:
            raw = json.loads(path.read_text(encoding="utf-8"))
            scene_id = raw.get("id", path.stem)
            category = raw.get("category", "uncategorized")

            bg = raw.get("background", {})
            bg_src = bg.get("src", "")
            background = BackgroundAssetOut(
                id=Path(bg_src).stem if bg_src else scene_id,
                src=bg_src,
                type=bg.get("type", "image"),
                thumb_src=bg.get("thumb_src"),
                loop=bg.get("loop", True),
                opacity=bg.get("opacity", 1.0),
                brightness=bg.get("brightness", 1.0),
                grayscale=bg.get("grayscale", 0.0),
                blur=bg.get("blur", 0.0),
                flip=bg.get("flip", False),
                blend_mode=bg.get("blend_mode", "normal"),
            )

            layers = [
                LayerAssetOut(
                    id=Path(lr.get("src", "")).stem or f"{scene_id}-layer-{i}",
                    src=lr.get("src", ""),
                    type=lr.get("type", "video"),
                    order=i,
                    loop=lr.get("loop", True),
                    opacity=lr.get("opacity", 1.0),
                    brightness=lr.get("brightness", 1.0),
                    grayscale=lr.get("grayscale", 0.0),
                    blur=lr.get("blur", 0.0),
                    flip=lr.get("flip", False),
                    blend_mode=lr.get("blend_mode", "normal"),
                )
                for i, lr in enumerate(raw.get("layers", []))
            ]

            scene = SceneOut(
                id=scene_id,
                label=raw.get("label", _label(scene_id)),
                background=background,
                layers=layers,
            )
            scenes.append(scene)
            cat_scenes.setdefault(category, []).append(
                SceneCategoryEntryOut(id=scene_id, label=scene.label)
            )

        except Exception as exc:
            logger.warning("Skipped scene %s — %s", path.name, exc)

    scene_categories = [
        SceneCategoryOut(
            id=cat,
            label=_label(cat),
            order=_order(cat),
            scenes=entries,
        )
        for cat, entries in sorted(cat_scenes.items(), key=lambda x: _order(x[0]))
    ]

    return scene_categories, scenes


def scan(scenes_dir: Path) -> AppData:
    """Scan R2 and local scene configs, return a fully populated AppData."""
    logger.info("Starting R2 scan...")
    keys = r2.list_keys()
    logger.info("Found %d objects in R2", len(keys))

    ambience_categories, ambiences = _scan_ambiences(keys)
    playlist_categories, playlists = _scan_playlists(keys)
    scene_categories, scenes = _scan_scenes(scenes_dir)

    logger.info(
        "Scan complete — %d ambience categories, %d ambiences, "
        "%d playlist categories, %d playlists, "
        "%d scene categories, %d scenes",
        len(ambience_categories), len(ambiences),
        len(playlist_categories), len(playlists),
        len(scene_categories), len(scenes),
    )

    return AppData(
        ambience_categories=ambience_categories,
        ambiences=ambiences,
        playlist_categories=playlist_categories,
        playlists=playlists,
        scene_categories=scene_categories,
        scenes=scenes,
        last_synced=datetime.now(timezone.utc).isoformat(),
    )

File: app/src/lib/server/scanner.ts
/**
 * R2 folder scanner. Builds in-memory AppData from the bucket folder structure.
 *
 * R2 folder conventions:
 *   ambiences/{category-slug}/{ambience-slug}.ogg
 *   ambiences/{category-slug}/cover.webp
 *   ambiences/{category-slug}/cover.thumb.webp
 *
 *   playlists/{category-slug}/{playlist-slug}/{track-slug}.ogg
 *   playlists/{category-slug}/{playlist-slug}/cover.webp
 *   playlists/{category-slug}/{playlist-slug}/cover.thumb.webp
 *
 *   scenes/{category-slug}/{scene-slug}.json
 *
 * Slugs (filename stems) are used as stable entity IDs and must be globally
 * unique within their type.
 */

import { S3Client, ListObjectsV2Command, GetObjectCommand } from "@aws-sdk/client-s3";
import { env } from "$env/dynamic/private";
import { PUBLIC_ASSETS_BASE } from "$env/static/public";
import type {
  Ambience,
  AmbienceCategory,
  AmbienceCategoryEntry,
} from "$lib/types/ambience";
import type {
  MusicTrack,
  Playlist,
  PlaylistCategory,
  PlaylistCategoryEntry,
} from "$lib/types/music";
import type {
  BlendMode,
  BackgroundAsset,
  LayerAsset,
  Scene,
  SceneCategory,
  SceneCategoryEntry,
} from "$lib/types/scene";
import type { AppData } from "$lib/types/appData";

/** Build a full CDN URL from a raw R2 object key. */
function assetUrl(src: string): string {
  return `${PUBLIC_ASSETS_BASE}/${src}`;
}

// ---------------------------------------------------------------------------
// Slug helpers
// ---------------------------------------------------------------------------

/** Strip leading numeric prefix and convert hyphens to title-cased words. */
function toLabel(slug: string): string {
  return slug
    .replace(/^\d+-/, "")
    .replace(/-/g, " ")
    .replace(/\b\w/g, (c) => c.toUpperCase());
}

/** Strip leading numeric prefix, keeping hyphens. Used for stable IDs. */
function toBase(slug: string): string {
  return slug.replace(/^\d+-/, "");
}

/** Return the leading numeric prefix of a slug as a sort key, 999 if absent. */
function toOrder(slug: string): number {
  const m = slug.match(/^(\d+)-/);
  return m ? parseInt(m[1], 10) : 999;
}

// ---------------------------------------------------------------------------
// R2 client
// ---------------------------------------------------------------------------

function createClient(): S3Client {
  return new S3Client({
    region: "auto",
    endpoint: `https://${env.R2_ACCOUNT_ID}.r2.cloudflarestorage.com`,
    credentials: {
      accessKeyId: env.R2_ACCESS_KEY,
      secretAccessKey: env.R2_SECRET_KEY,
    },
  });
}

/** Return every object key in the bucket, handling pagination. */
async function listR2Keys(client: S3Client): Promise<string[]> {
  const keys: string[] = [];
  let continuationToken: string | undefined;
  do {
    const res = await client.send(
      new ListObjectsV2Command({
        Bucket: env.R2_BUCKET,
        ContinuationToken: continuationToken,
      }),
    );
    for (const obj of res.Contents ?? []) if (obj.Key) keys.push(obj.Key);
    continuationToken = res.NextContinuationToken;
  } while (continuationToken);

  return keys;
}

// ---------------------------------------------------------------------------
// Ambiences
// ---------------------------------------------------------------------------

/** Build ambience categories and a flat ambience list from R2 keys. */
function scanAmbiences(keys: string[]): {
  categories: AmbienceCategory[];
  ambiences: Ambience[];
} {
  const cats = new Map<string, { src: string; thumb_src: string | null }>();
  const ambiences = new Map<string, Ambience>();
  const catAmbs = new Map<string, string[]>();

  for (const key of keys) {
    const parts = key.split("/");
    if (parts.length !== 3 || parts[0] !== "ambiences") continue;
    const [, catSlug, filename] = parts;

    if (!cats.has(catSlug)) cats.set(catSlug, { src: "", thumb_src: null });

    if (filename === "cover.webp") {
      cats.get(catSlug)!.src = key;
    } else if (filename === "cover.thumb.webp") {
      cats.get(catSlug)!.thumb_src = key;
    } else if (filename.endsWith(".ogg")) {
      const ambSlug = filename.slice(0, -4);
      const id = `${toBase(catSlug)}.${ambSlug}`;
      if (ambiences.has(id)) {
        console.warn(
          `Duplicate ambience '${ambSlug}' in '${catSlug}' — skipping`,
        );
        continue;
      }
      ambiences.set(id, {
        id,
        label: toLabel(ambSlug),
        loop: true,
        src: key,
        url: assetUrl(key),
      });
      if (!catAmbs.has(catSlug)) catAmbs.set(catSlug, []);
      catAmbs.get(catSlug)!.push(ambSlug);
    }
  }

  const categories: AmbienceCategory[] = [...cats.entries()]
    .sort((a, b) => toOrder(a[0]) - toOrder(b[0]))
    .map(([catSlug, meta]) => {
      const catBase = toBase(catSlug);
      const entries: AmbienceCategoryEntry[] = (catAmbs.get(catSlug) ?? [])
        .sort()
        .map((slug) => ({ id: `${catBase}.${slug}`, label: toLabel(slug) }));
      return {
        id: toBase(catSlug),
        label: toLabel(catSlug),
        src: meta.src,
        thumb_src: meta.thumb_src,
        url: assetUrl(meta.src),
        thumb_url: meta.thumb_src ? assetUrl(meta.thumb_src) : null,
        order: toOrder(catSlug),
        ambiences: entries,
      };
    });

  return { categories, ambiences: [...ambiences.values()] };
}

// ---------------------------------------------------------------------------
// Playlists
// ---------------------------------------------------------------------------

/** Build playlist categories and full playlist objects from R2 keys. */
function scanPlaylists(keys: string[]): {
  categories: PlaylistCategory[];
  playlists: Playlist[];
} {
  const cats = new Set<string>();
  const plData = new Map<
    string,
    { src: string; thumb_src: string | null; tracks: MusicTrack[] }
  >();

  for (const key of keys) {
    const parts = key.split("/");
    if (parts.length !== 4 || parts[0] !== "playlists") continue;
    const [, catSlug, plSlug, filename] = parts;

    cats.add(catSlug);
    const plKey = `${catSlug}/${plSlug}`;
    if (!plData.has(plKey))
      plData.set(plKey, { src: "", thumb_src: null, tracks: [] });

    const pl = plData.get(plKey)!;
    if (filename === "cover.webp") pl.src = key;
    else if (filename === "cover.thumb.webp") pl.thumb_src = key;
    else if (filename.endsWith(".ogg"))
      pl.tracks.push({ id: filename.slice(0, -4), src: key });
  }

  for (const pl of plData.values())
    pl.tracks.sort((a, b) => a.src.localeCompare(b.src));

  const catEntries = new Map<string, PlaylistCategoryEntry[]>();
  for (const cat of cats) catEntries.set(cat, []);

  const playlists: Playlist[] = [];
  for (const [plKey, meta] of [...plData.entries()].sort()) {
    const slash = plKey.indexOf("/");
    const catSlug = plKey.slice(0, slash);
    const plSlug = plKey.slice(slash + 1);
    playlists.push({
      id: plSlug,
      label: toLabel(plSlug),
      src: meta.src,
      thumb_src: meta.thumb_src,
      url: assetUrl(meta.src),
      thumb_url: meta.thumb_src ? assetUrl(meta.thumb_src) : null,
      tracks: meta.tracks.map((t) => ({ ...t, url: assetUrl(t.src) })),
    });
    catEntries.get(catSlug)!.push({ id: plSlug, label: toLabel(plSlug) });
  }

  const categories: PlaylistCategory[] = [...cats]
    .sort((a, b) => toOrder(a) - toOrder(b))
    .map((catSlug) => ({
      id: toBase(catSlug),
      label: toLabel(catSlug),
      order: toOrder(catSlug),
      playlists: catEntries.get(catSlug) ?? [],
    }));

  return { categories, playlists };
}

// ---------------------------------------------------------------------------
// Scenes — fetched from R2 at scenes/{category-slug}/{scene-slug}.json
// ---------------------------------------------------------------------------

/** Fetch and parse scene JSON configs from R2, building scene + category structures. */
async function scanScenes(
  keys: string[],
  client: S3Client,
): Promise<{ categories: SceneCategory[]; scenes: Scene[] }> {
  const sceneKeys = keys.filter((k) => {
    const parts = k.split("/");
    return (
      parts.length === 3 && parts[0] === "scenes" && parts[2].endsWith(".json")
    );
  });

  const scenes: Scene[] = [];
  const catScenes = new Map<string, SceneCategoryEntry[]>();

  for (const key of sceneKeys) {
    const [, catSlug, filename] = key.split("/");
    const sceneSlug = filename.slice(0, -5);

    try {
      const res = await client.send(
        new GetObjectCommand({ Bucket: env.R2_BUCKET, Key: key }),
      );
      const body = await res.Body?.transformToString();
      if (!body) continue;

      const raw = JSON.parse(body) as Record<string, unknown>;
      const sceneId = (raw.id as string | undefined) ?? sceneSlug;

      const bg = (raw.background as Record<string, unknown> | undefined) ?? {};
      const bgSrc = (bg.src as string | undefined) ?? "";
      const bgThumbSrc = (bg.thumb_src as string | null | undefined) ?? null;

      const background: BackgroundAsset = {
        id: bgSrc
          ? bgSrc.split("/").pop()!.replace(/\.[^.]+$/, "")
          : sceneId,
        src: bgSrc,
        type: (bg.type as "image" | "video" | undefined) ?? "image",
        thumb_src: bgThumbSrc,
        url: bgSrc ? assetUrl(bgSrc) : undefined,
        thumb_url: bgThumbSrc ? assetUrl(bgThumbSrc) : null,
        loop: (bg.loop as boolean | undefined) ?? true,
        opacity: (bg.opacity as number | undefined) ?? 1.0,
        brightness: (bg.brightness as number | undefined) ?? 1.0,
        grayscale: (bg.grayscale as number | undefined) ?? 0.0,
        blur: (bg.blur as number | undefined) ?? 0.0,
        flip: (bg.flip as boolean | undefined) ?? false,
        blend_mode: (bg.blend_mode as BlendMode | undefined) ?? "normal",
      };

      const layers: LayerAsset[] = (
        (raw.layers as unknown[] | undefined) ?? []
      ).map((lr, i) => {
        const l = lr as Record<string, unknown>;
        const lSrc = (l.src as string | undefined) ?? "";
        return {
          id: lSrc
            ? lSrc.split("/").pop()!.replace(/\.[^.]+$/, "")
            : `${sceneId}-layer-${i}`,
          src: lSrc,
          url: lSrc ? assetUrl(lSrc) : undefined,
          type: (l.type as "image" | "video" | undefined) ?? "video",
          order: i,
          loop: (l.loop as boolean | undefined) ?? true,
          opacity: (l.opacity as number | undefined) ?? 1.0,
          brightness: (l.brightness as number | undefined) ?? 1.0,
          grayscale: (l.grayscale as number | undefined) ?? 0.0,
          blur: (l.blur as number | undefined) ?? 0.0,
          flip: (l.flip as boolean | undefined) ?? false,
          blend_mode: (l.blend_mode as BlendMode | undefined) ?? "normal",
        };
      });

      const label = (raw.label as string | undefined) ?? toLabel(sceneSlug);
      scenes.push({ id: sceneId, label, background, layers });

      if (!catScenes.has(catSlug)) catScenes.set(catSlug, []);
      catScenes.get(catSlug)!.push({ id: sceneId, label });
    } catch (err) {
      console.warn(`Skipped scene ${key}:`, err);
    }
  }

  const categories: SceneCategory[] = [...catScenes.entries()]
    .sort((a, b) => toOrder(a[0]) - toOrder(b[0]))
    .map(([catSlug, entries]) => ({
      id: toBase(catSlug),
      label: toLabel(catSlug),
      order: toOrder(catSlug),
      scenes: entries,
    }));

  return { categories, scenes };
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

/** Scan R2 for all assets and scene configs, return a fully populated AppData snapshot. */
export async function scan(): Promise<AppData> {
  console.log("Starting R2 scan...");
  const client = createClient();
  const keys = await listR2Keys(client);
  console.log(`Found ${keys.length} objects in R2`);

  const { categories: ambienceCategories, ambiences } = scanAmbiences(keys);
  const { categories: playlistCategories, playlists } = scanPlaylists(keys);
  const { categories: sceneCategories, scenes } = await scanScenes(keys, client);

  console.log(
    `Scan complete — ${ambienceCategories.length} ambience categories, ${ambiences.length} ambiences, ` +
      `${playlistCategories.length} playlist categories, ${playlists.length} playlists, ` +
      `${sceneCategories.length} scene categories, ${scenes.length} scenes`,
  );

  return {
    ambience_categories: ambienceCategories,
    ambiences,
    playlist_categories: playlistCategories,
    playlists,
    scene_categories: sceneCategories,
    scenes,
    last_synced: new Date().toISOString(),
  };
}
